src/sensai/clustering/base/clustering.py

```diff
@@ -0,0 +1,196 @@
+            # logging process through the loop
+            percentageGenerated = int(100 * i / self.numClusters)
+            if percentageGenerated == percentageToLog:
+                log.info(f"Processed {percentageToLog}% of clusters")
+                percentageToLog += 5
+
+            cluster = self.getCluster(clusterId)
+            if condition is None or condition(cluster):
+                yield cluster
+
+    def noiseCluster(self):
+        if self.noiseLabel is None:
+            raise NotImplementedError(f"The algorithm {self} does not provide a noise cluster")
+        return self.getCluster(self.noiseLabel)
+
+    def summaryDF(self, condition: Callable[[Cluster], bool] = None):
+        """
+        Data frame containing coarse information about the clusters
+
+        :param condition: if provided, only clusters fulfilling the condition will be included
+        :return: pandas DataFrame
+        """
+        summary_dicts = [cluster.summaryDict() for cluster in self.clusters(condition=condition)]
+        return pd.DataFrame(summary_dicts).set_index("identifier", drop=True)
+
+    def fit(self, data: np.ndarray) -> None:
+        log.info(f"Fitting {self} to {len(data)} coordinate datapoints.")
+        self._fit(data)
+        self._datapoints = data
+        self._labels = self.getLabels()
+        if len(self._labels) != len(data):
+            raise Exception(f"Number of labels does not match number of datapoints")
+        self._clusterIdentifiers = set(self._labels)
+        log.info(f"{self} found {self.numClusters} clusters")
+
+    @property
+    def isFitted(self):
+        return self._datapoints is not None
+
+    @property
+    def datapoints(self) -> np.ndarray:
+        assert self.isFitted
+        return self._datapoints
+
+    def getLabels(self) -> np.ndarray:
+        assert self.isFitted
+        labels = self._getLabels()
+        # Relabel clusters that do not fulfill size bounds as noise
+        if self.minClusterSize != -np.inf or self.maxClusterSize != np.inf:
+            for clusterId, clusterSize in zip(*np.unique(labels, return_counts=True)):
+                if not self.minClusterSize <= clusterSize <= self.maxClusterSize:
+                    labels[labels == clusterId] = self.noiseLabel
+        return labels
+
+    @property
+    def clusterIdentifiers(self) -> Set[int]:
+        assert self.isFitted
+        return self._clusterIdentifiers
+
+    # unfortunately, there seems to be no way to annotate the return type correctly
+    # https://github.com/python/mypy/issues/3993
+    def getCluster(self, clusterId: int) -> Cluster:
+        if clusterId not in self.getLabels():
+            raise KeyError(f"no cluster for id {clusterId}")
+        return self.Cluster(self.datapoints[self.getLabels() == clusterId], identifier=clusterId)
+
+    @property
+    def numClusters(self) -> int:
+        return len(self.clusterIdentifiers.difference({self.noiseLabel}))
```

Comment:
Since we use this difference construct several times, we might want to consider leaving the noise label out or keeping two lists of identifiers. Especially since you used the property annotation here, I expect this to be O(1), but that's far from the case. I highly discourage the use of the property annotation particularly in such situations. (Basically, I think it should only be used as a tool to establish backward compatibility - when a property has been removed but can be functionally determined by other properties - and not as a general-purpose pattern.)

_General observation_: Your implementations don't put much emphasis on computational efficiency. We can certainly apply such a style in our own project-specific code, but we should pay more attention to efficiency in the library (whenever it doesn't lead to overly complex code), as inefficiencies could lead to severe runtime penalties down the line.

Reply:
This one I can get in O(1) simply by substracting 1 when the noise label is not None. That should be good enough, right?